### PR TITLE
[feat][fix] ShardedDDP - better tests, removing broken reduce fp16 (to be reintroduced properly)

### LINKED ISSUE #517: [ShardedDDP] Grad norm disparity with DDP when reduce buckets are enabled
## 🐛 Bug
The grad norms become false when bucketing is enabled with ShardedDDP, spotted by Weiyi (@zhengwy888), many thanks !

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
- Fixes #517 as far as I can tell
- Removes the reduce fp16 option from ShardedDDP, which was broken, other PR incoming to re-implement it properly
- Adds a couple of cases in the DDP parity unit test, among which
-- check that the grads are the same
-- handle AMP on/off
-- handle Gloo/NCCL

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 1]
```diff
@@ -300,8 +290,8 @@ def reduce(self) -> None:
         # Trigger all the current BW hooks
         _ = map(lambda x: x(), self._grad_accs)
 
-        # Make sure that all the futures are consumed
-        self._consume_work_handles()
+        # Make sure that the buckets are flushed and all the futures are consumed
+        self._flush_reduce_calls()
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:294]: manual reduce case, spotted by the unit test, this was not flushing the buckets (so manual reduce + unused params would not be correct)

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 2]
```diff
@@ -413,22 +403,16 @@ def reduce(*_: Any) -> None:
...
-                        else:
-                            assert param.grad is not None
-                            param.grad.data = param.grad.data.to(dtype=param.dtype)
 
                     # Async reduce for this buffer, log the future
                     self._work_handles.append(
                         Workhandle(
                             handle=dist.reduce(
-                                tensor=param.grad.data,
+                                tensor=param.grad,
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:415]: to my understanding this does not change anything -unit test agrees-, just better not to use .data when not needed I think

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 3]
```diff
@@ -461,6 +445,12 @@ def reduce(*_: Any) -> None:
                         # Normalize the bucket in one go
                         bucket.buffer.mul_(self.world_size_scaling)
 
+                        def cleanup() -> None:
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:448]: this should not be strictly needed if users
- use OSS.clip_grad_norm
- use grad accumulation with the no_sync() context
but in case one of these was not done, it would lead to gradients being both in the original buckets and in the reduced buckets, which can lead to errors. This is then arguably cleaner, reduced buckets are empty

File: fairscale/optim/oss.py [hunk 4]
```diff
@@ -247,6 +248,7 @@ def clip_grad_norm(
         Arguments:
             max_norm (float or int): max norm of the gradients
             norm_type (float or int): type of the used p-norm. Can be ``'inf'`` for infinity norm.
+            skip_nan_grads (bool): auto-skip scaling if some grads are NaNs (which can happen with Pytorch AMP, in that case the model update is skipped)
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:251]: if NaN grads are ok (basically if you're using torch AMP), then this can be passed

File: fairscale/optim/oss.py [hunk 5]
```diff
@@ -286,10 +288,20 @@ def clip_grad_norm(
             total_norm = total_norm ** (1.0 / norm_type)
 
         clip_coef = torch.tensor(max_norm, dtype=total_norm.dtype, device=total_norm.device) / (total_norm + 1e-6)
+
+        if skip_nan_grads and torch.isnan(clip_coef):
+            return total_norm
+
+        assert not torch.isnan(clip_coef), (
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:295]: by default we assume that normalizing by a NaN is most probably the sign that something is broken, better signal it early

File: fairscale/optim/utils.py [hunk 6]
```diff
@@ -120,5 +120,6 @@ def calc_grad_norm(parameters: List[torch.nn.Parameter], p: float) -> torch.Tens
     if p == inf:
         local_norm = max(par.grad.detach().abs().max() for par in parameters)  # type: ignore
     else:
-        local_norm = torch.norm(torch.stack([torch.norm(par.grad.detach(), p) for par in parameters]), p)  # type: ignore
+        # Compute the norm in full precision no matter what
+        local_norm = torch.norm(torch.stack([torch.norm(par.grad.detach(), p, dtype=torch.float32) for par in parameters]), p).to(dtype=parameters[0].dtype)  # type: ignore
```
Comment (blefaudeux) [inline fairscale/optim/utils.py:124]: fixed in https://github.com/facebookresearch/fairscale/pull/520, in some cases the sum on fp16 could overflow. The cost is negligible basically, and this minimizes the computations done in fp16

File: tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py [hunk 7]
```diff
@@ -136,35 +138,42 @@ def step():
...
+                    _scaler.step(_optimizer)
+                    _scaler.update()
 
             check_same_model_params(sharded_ddp_model, ddp_model, f"Rank: {rank} - Step {i} broke")
 
+            # Check that the two grad norm are equivalent
+            # NOTE: The grads can occasionally be NaNs, the scaler will skip the step in that case
+            # This is not ShardedDDP specific. If the grads are not NaN for DDP then they should also
+            # be valid for ShardedDDP
+            if clip_grad_norm:
```
Comment (blefaudeux) [inline tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py:169]: *new* check that clip_grad_norm is also at parity in between DDP and ShardedDDP

File: tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py [hunk 8]
```diff
@@ -201,10 +204,11 @@ def closure_sharded(input_tensor=input_tensor):
 @pytest.mark.parametrize("reduce_buffer_size", [0, 2 ** 20])
 @pytest.mark.parametrize("grad_accumulation", [True, False])
 @pytest.mark.parametrize("change_train_graph", [True, False])
-@pytest.mark.parametrize("fp16_reduction", _test_fp16_reduction)
-def test_ddp_parity(reduce_buffer_size, grad_accumulation, change_train_graph, fp16_reduction):
+@pytest.mark.parametrize("clip_grad_norm", [True, False])
```
Comment (blefaudeux) [inline tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py:207]: bigger test matrix, since #517 showed that the current test matrix was not enough..

**Comment by blefaudeux** [comment]
no NaNs, but the different in the grad norm is noticeably bigger with torch 1.6 and 1.7, worth investigating I think

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 9]
```diff
@@ -413,22 +404,19 @@ def reduce(*_: Any) -> None:
...
                     # Future work includes clearing up the buffer if possible
                     def cleanup() -> None:
                         if dst_rank != self.global_rank:
                             param.grad = None
-                        else:
-                            assert param.grad is not None
-                            param.grad.data = param.grad.data.to(dtype=param.dtype)
+
+                    if self.backend == dist.Backend.NCCL:
+                        torch.cuda.synchronize()
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:413]: doing that completely tanks the perf, but the results are not correct if I remove just this line (same for the 2 other same lines below).. cc @ngimel @mrshenli, any other way to make sure that the NCCL stream(s) wait for a given event on the current one ?

**Comment by blefaudeux** [comment]
changing more things to get to the bottom of all the cuda sync issues
